exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/TopNBatch.java

```diff
@@ -162,56 +173,67 @@ public void buildSchema() throws SchemaChangeException {
         return;
       case NONE:
         state = BatchState.DONE;
+      case EMIT:
+        throw new IllegalStateException("Unexpected EMIT outcome received in buildSchema phase");
       default:
-        return;
+        throw new IllegalStateException("Unexpected outcome received in buildSchema phase");
     }
   }
 
   @Override
   public IterOutcome innerNext() {
     recordCount = 0;
     if (state == BatchState.DONE) {
-      return IterOutcome.NONE;
+      return NONE;
     }
-    if (schema != null) {
-      if (getSelectionVector4().next()) {
+
+    // If both schema and priorityQueue are non-null and priority queue is not reset, that means we still have data
+    // to be sent downstream for the current record boundary
+    if (schema != null && priorityQueue != null && priorityQueue.isInitialized()) {
+      if (sv4.next()) {
         recordCount = sv4.getCount();
-        return IterOutcome.OK;
+        container.setRecordCount(recordCount);
       } else {
         recordCount = 0;
-        return IterOutcome.NONE;
+        container.setRecordCount(0);
       }
+      return getFinalOutcome();
     }
 
     try{
       outer: while (true) {
         Stopwatch watch = Stopwatch.createStarted();
-        IterOutcome upstream;
         if (first) {
-          upstream = IterOutcome.OK_NEW_SCHEMA;
+          laskKnownOutcome = IterOutcome.OK_NEW_SCHEMA;
           first = false;
         } else {
-          upstream = next(incoming);
+          laskKnownOutcome = next(incoming);
         }
-        if (upstream == IterOutcome.OK && schema == null) {
-          upstream = IterOutcome.OK_NEW_SCHEMA;
+        if (laskKnownOutcome == OK && schema == null) {
+          laskKnownOutcome = IterOutcome.OK_NEW_SCHEMA;
           container.clear();
         }
         logger.debug("Took {} us to get next", watch.elapsed(TimeUnit.MICROSECONDS));
-        switch (upstream) {
+        switch (laskKnownOutcome) {
         case NONE:
           break outer;
         case NOT_YET:
           throw new UnsupportedOperationException();
         case OUT_OF_MEMORY:
         case STOP:
-          return upstream;
+          return laskKnownOutcome;
         case OK_NEW_SCHEMA:
           // only change in the case that the schema truly changes.  Artificial schema changes are ignored.
+          // schema change handling in case when EMIT is also seen is same as without EMIT. i.e. only if union type
+          // is enabled it will be handled.
+          container.clear();
+          firstBatchForSchema = true;
           if (!incoming.getSchema().equals(schema)) {
```

Comment:
equals() or isEquivalent() ?

Reply:
I am keeping it as is for now. Don't want to change any behavior related to schema change handling.